### PR TITLE
[MODULAR] NIF Rewards Points

### PR DESCRIPTION
**Pull request by softcerv**
## About The Pull Request
Adds rewards points to NIFs, a system that aims to make rebuying NIFSofts every shift less of a pain. 

When buying an eligible NIFSoft from the NIFSoft catalog, a fraction of the buy price will be reimbursed as reward points, these rewards points can be used to purchase additional NIFSofts in place of credits. These rewards points carry between rounds.

The default conversion rate is 50% of the NIFSofts cost. Additionally, any NIFSofts installed from disks do not give rewards points.

## How This Contributes To The Skyrat Roleplay Experience
Buying NIFSofts each round is a pain, especially with the more expensive NIFSofts like Hivemind. This should hopefully make that cycle less of a pain.

## Proof of Testing

<details>
<summary>Screenshots/Videos</summary>

image

image

image
  
</details>

## Changelog

:cl:
add: added reward points to NIFs, an alternative currency given from buying NIFSofts that can be used to buy more NIFSofts.
/:cl:

### CODE REVIEW

**Comment by softcerv** [comment]
going to keep this on a draft while I fix weakrefs

File: modular_skyrat/modules/modular_implants/code/nifsoft_catalog.dm [hunk 1]
```diff
@@ -32,7 +32,15 @@ GLOBAL_LIST_INIT(purchasable_nifsofts, list(
 
 	paying_account = computer.computer_id_slot?.registered_account || null
 	data["paying_account"] = paying_account
+	data["current_balance"] = computer.computer_id_slot?.registered_account?.account_balance || null
 
+	var/rewards_points = 0
+	var/obj/item/organ/internal/cyberimp/brain/nif/buyer_nif = target_nif
```
Comment (Zonespace27) [inline modular_skyrat/modules/modular_implants/code/nifsoft_catalog.dm]:
```suggestion
	var/obj/item/organ/internal/cyberimp/brain/nif/buyer_nif = target_nif.resolve()
```

File: modular_skyrat/modules/modular_implants/code/nifsoft_catalog.dm [hunk 2]
```diff
@@ -42,7 +50,7 @@ GLOBAL_LIST_INIT(purchasable_nifsofts, list(
 	var/mob/living/carbon/human/nif_user = user
 	if(nif_user)
 		var/obj/item/organ/internal/cyberimp/brain/nif/user_nif = nif_user.getorgan(/obj/item/organ/internal/cyberimp/brain/nif)
-		target_nif = user_nif || null
+		target_nif = user_nif
```
Comment (Zonespace27) [inline modular_skyrat/modules/modular_implants/code/nifsoft_catalog.dm]:
```suggestion
		target_nif = weakref(user_nif)
```
I *think* that's the syntax?

File: modular_skyrat/modules/modular_implants/code/nifsoft_catalog.dm [hunk 3]
```diff
@@ -80,12 +90,19 @@ GLOBAL_LIST_INIT(purchasable_nifsofts, list(
 				return FALSE
 
 			var/amount_to_charge = (params["product_cost"])
+			var/rewards_purchase = (params["rewards_purchase"])
+			var/obj/item/organ/internal/cyberimp/brain/nif/buyer_nif = target_nif
```
Comment (Zonespace27) [inline modular_skyrat/modules/modular_implants/code/nifsoft_catalog.dm]:
```suggestion
			var/obj/item/organ/internal/cyberimp/brain/nif/buyer_nif = target_nif.resolve()
```

File: modular_skyrat/modules/modular_implants/code/nifsofts.dm [hunk 4]
```diff
@@ -39,9 +39,17 @@
 	///What NIF models can this software be installed on?
 	var/list/compatible_nifs = list(/obj/item/organ/internal/cyberimp/brain/nif)
 
-/datum/nifsoft/New(obj/item/organ/internal/cyberimp/brain/nif/recepient_nif)
+	/// How much of the NIFSoft's purchase price is payed out as reward points, if any?
```
Comment (Zonespace27) [inline modular_skyrat/modules/modular_implants/code/nifsofts.dm]:
```suggestion
	/// How much of the NIFSoft's purchase price is paid out as reward points, if any?
```

**Review by Zonespace27 — CHANGES_REQUESTED** [review]
Make it clear that these are cross-round; currently there's no indication as such

**Comment by YakumoChen** [comment]
What's the point of giving some of these a price beyond "roundstart buyable" if you're going to add a system that makes them feasibly roundstart buyable?

**Comment by YakumoChen** [comment]
Are points used in the purchase or is it just a flat scaling discount with purchases?

**Comment by YakumoChen** [comment]
Is the 50% conversion for gaining points or spending them? Are they 1:1 with credits?

**Comment by YakumoChen** [comment]
If you buy using points does the catalogue still try and give you points for the purchase? It should be made clear only credit purchases reward points.

**Comment by softcerv** [comment]
> What's the point of giving some of these a price beyond "roundstart buyable" if you're going to add a system that makes them feasibly roundstart buyable?

This is only 50% of the money that you spend. You would be able to buy the same NIFSoft with the points roundstart once every three rounds assuming you bought the same NIFSoft the previous two rounds. The price is still important here, this just makes it somewhat less of an issue.

> Are points used in the purchase or is it just a flat scaling discount with purchases?

As shown by the screenshots I posted under the proof of testing section, they are an alternate way to pay.

> If you buy using points does the catalogue still try and give you points for the purchase? It should be made clear only credit purchases reward points.

No, rewards points are not given when purchasing items with rewards points. This is similar to how a lot of rewards systems work in real life.

Next time it would help if you posted all of these as one single comment.

**Comment by YakumoChen** [comment]
Is this exploitable in any way? If I came across 20,000 credits and bought and deleted NIF softs until I had 10,000 points, I'd be set for weeks.

Since this is persistent, there should probably be a pretty low cap on points to prevent abuse and encourage people to use the points. Something like 2000.

**Comment by softcerv** [comment]
> Is this exploitable in any way? If I came across 20,000 credits and bought and deleted NIF softs until I had 10,000 points, I'd be set for weeks.
> 
> Since this is persistent, there should probably be a pretty low cap on points to prevent abuse and encourage people to use the points. Something like 2000.

You would need to repeatedly uninstall and install NIFSofts for a mind numbing amount of time to do that. I am honestly not sure of how real of an issue this is going to be.

**Comment by YakumoChen** [comment]
Mind numbing, but better safe than sorry. If someone is hitting the cap and not spending the points, they don't really have a reason to complain 

I wouldn't like having someone spending a round farming alt-credits and getting an absurd amount that persists.

**Review by Gandalf2k15 — APPROVED** [review]

**Pull request merged**